crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+                let (count, i) = match alias {
+                    Alias::Outer {
+                        count,
+                        i,
+                        kind: OuterAliasKind::Type,
+                    } => (count, i),
+                    _ => unreachable!(),
+                };
+                let ty = self.outer_type(count, i).clone();
+                self.current_type_scope_mut().push(ty);
+                Ok(InstanceTypeDef::Alias(alias))
+            }
+
+            // Type definition.
+            _ => {
+                let ty = self.arbitrary_type(u)?;
+                Ok(InstanceTypeDef::Type(ty))
+            }
+        }
+    }
+
+    fn arbitrary_outer_type_alias(&mut self, u: &mut Unstructured<'a>) -> Result<Alias> {
+        let non_empty_types_scopes: Vec<_> = self
+            .types_scopes
+            .iter()
+            .rev()
+            .enumerate()
+            .filter(|(_, scope)| !scope.is_empty())
+            .collect();
+        assert!(
+            !non_empty_types_scopes.is_empty(),
+            "precondition: there are non-empty types scopes"
+        );
+
+        let (count, scope) = u.choose(&non_empty_types_scopes)?;
+        let count = u32::try_from(*count).unwrap();
+        assert!(!scope.is_empty());
+
+        let max_type_in_scope = u32::try_from(scope.len() - 1).unwrap();
+        let i = u.int_in_range(0..=max_type_in_scope)?;
+
+        Ok(Alias::Outer {
+            count,
+            i,
+            kind: OuterAliasKind::Type,
+        })
+    }
+
+    fn arbitrary_func_type(&mut self, u: &mut Unstructured<'a>) -> Result<FuncType> {
+        let mut params = vec![];
+        let mut param_names = HashSet::new();
+        arbitrary_loop(u, 0, 20, |u| {
+            params.push(self.arbitrary_named_type(u, &mut param_names)?);
+            Ok(true)
+        })?;
+
+        let result = self.arbitrary_value_type(u)?;
+
+        Ok(FuncType { params, result })
+    }
+
+    fn arbitrary_value_type(&mut self, u: &mut Unstructured<'a>) -> Result<ValueType> {
+        let num_choices = if self.current_type_scope().is_empty() {
+            13
+        } else {
+            14
+        };
+        match u.int_in_range(0..=num_choices)? {
+            0 => Ok(ValueType::Unit),
+            1 => Ok(ValueType::Bool),
+            2 => Ok(ValueType::S8),
+            3 => Ok(ValueType::U8),
+            4 => Ok(ValueType::S16),
+            5 => Ok(ValueType::U16),
+            6 => Ok(ValueType::S32),
+            7 => Ok(ValueType::U32),
+            8 => Ok(ValueType::S64),
+            9 => Ok(ValueType::U64),
+            10 => Ok(ValueType::F32),
+            11 => Ok(ValueType::F64),
+            12 => Ok(ValueType::Char),
+            13 => Ok(ValueType::String),
+            14 => {
+                let max_type = u32::try_from(self.current_type_scope().len() - 1).unwrap();
+                Ok(ValueType::Indexed(u.int_in_range(0..=max_type)?))
```

Comment:
Naively this feels like it might generate an invalid module, because I think types can point to things like functions and modules, right? (and we can't have value types for those)

Reply:
I don't think that types can point to functions and modules? These types, here and elsewhere, are transcribed from the grammar in the explainer and binary documents.